tools/fuzzer/platform/imx8-host.c

```diff
@@ -0,0 +1,559 @@
+#define ADSP_SRAM_EXCEPT_SIZE       0x800
+#define ADSP_SRAM_EXCEPT_OFFSET     (ADSP_SRAM_DEBUG_OFFSET \
+				     + ADSP_SRAM_DEBUG_SIZE)
+
+#define ADSP_SRAM_STREAM_BASE       (ADSP_SRAM_EXCEPT_BASE \
+				     + ADSP_SRAM_EXCEPT_SIZE)
+#define ADSP_SRAM_STREAM_SIZE       0x1000
+#define ADSP_SRAM_STREAM_OFFSET     (ADSP_SRAM_EXCEPT_OFFSET \
+				     + ADSP_SRAM_EXCEPT_SIZE)
+
+#define ADSP_SRAM_TRACE_BASE        (ADSP_SRAM_STREAM_BASE \
+				     + ADSP_SRAM_STREAM_SIZE)
+#define ADSP_SRAM_TRACE_SIZE        0x1000
+#define ADSP_SRAM_TRACE_OFFSET      (ADSP_SRAM_STREAM_OFFSET \
+				     + ADSP_SRAM_STREAM_SIZE)
+
+#define ADSP_IMX8_DSP_MAILBOX_SIZE  (ADSP_SRAM_INBOX_SIZE \
+				     + ADSP_SRAM_OUTBOX_SIZE \
+				     + ADSP_SRAM_DEBUG_SIZE \
+				     + ADSP_SRAM_EXCEPT_SIZE \
+				     + ADSP_SRAM_STREAM_SIZE \
+				     + ADSP_SRAM_TRACE_SIZE)
+#define ADSP_IMX8_IRAM_SIZE          0x8000
+#define ADSP_IMX8_DRAM_SIZE          0x8000
+
+#define ADSP_IMX8_DSP_MU_SIZE		0x10000
+#define ADSP_IMX8_DSP_MU13B_BASE	0x5D310000
+#define ADSP_IMX8_DSP_MU13A_BASE	0x5D280000
+
+#define ADSP_MAILBOX_SIZE		0x1000
+
+/* TODO get from driver. */
+#define IMX8_PANIC_OFFSET(x)	(x)
+
+struct imx8_data {
+	void *bar[MAX_BAR_COUNT];
+	struct mailbox host_box;
+	struct mailbox dsp_box;
+	int boot_complete;
+	pthread_mutex_t mutex;
+};
+
+/* Platform host description taken from Qemu
+ * - mapped to BAR 0, 1, 2, 3
+ */
+static struct fuzzer_mem_desc imx8_mem[] = {
+	{.name = "iram", .base = ADSP_IMX8_HOST_IRAM_BASE,
+	 .size = ADSP_IMX8_IRAM_SIZE},
+	{.name = "dram", .base = ADSP_IMX8_HOST_DRAM_BASE,
+	 .size = ADSP_IMX8_DRAM_SIZE},
+	{.name = "sdram0", .base = SDRAM0_BASE,
+	 .size = SDRAM1_SIZE},
+	{.name = "sdram1", .base = SDRAM1_BASE,
+	 .size = SDRAM1_SIZE},
+};
+
+/* mapped to BAR 4, 5, 6 */
+static struct fuzzer_reg_space imx8_io[] = {
+	{ .name = "mu_13a",
+	  .desc = {.base = ADSP_IMX8_DSP_MU13A_BASE,
+		   .size = ADSP_IMX8_DSP_MU_SIZE},},
+	{ .name = "mu_13b",
+	  .desc = {.base = ADSP_IMX8_DSP_MU13B_BASE,
+		   .size = ADSP_IMX8_DSP_MU_SIZE},},
+	{ .name = "mbox",
+	  .desc = {.base = ADSP_IMX8_DSP_MAILBOX_BASE,
+		   .size = ADSP_IMX8_DSP_MAILBOX_SIZE},},
+};
+
+#define IMX8_MU13_A_BAR		4
+#define IMX8_MU13_B_BAR		5
+#define IMX8_MBOX_BAR		6
+
+/*
+ * Platform support for i.MX8.
+ *
+ * The IPC portions below are copied and pasted from the SOF driver with some
+ * modification for data structure and printing.
+ *
+ * The "driver" code below no longer writes directly to the HW but writes
+ * to the virtual HW as exported by qemu as Posix SHM and message queues.
+ *
+ * Register IO and mailbox IO is performed using shared memory regions between
+ * fuzzer and qemu.
+ *
+ * IRQs are send using message queues between fuzzer and qemu.
+ *
+ * SHM and message queues can be inspected from the cmd line by using
+ * "less -C" on /dev/shm/name and /dev/mqueue/name
+ */
+
+static uint64_t imx8_fixup_side_b(struct fuzz *fuzzer, unsigned int bar,
```

Comment:
I think this function might be worth having a comment itself (clear any pending interrupts if you just sent something?)